doc: gs: add debugging section to Testing an app (#7973)
nrfconnect/sdk-nrf: doc/nrf/gs_testing.rst

#. Select the appropriate CPU for debugging in each session, corresponding to the application core and the network core of your SoC, respectively.
#. Once both sessions are established, execute the code on the application core.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
#. Start code execution on the network core in the other debug session.

If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.

Debugging secure and non-secure domains
=======================================

When using an ``_ns`` build target, by default you can only debug firmware in the non-secure part of the application core firmware.

To debug firmware running in the secure domain, you need to build Trusted Firmware-M with debug symbols enabled and load the symbols during the debugging session.
To build Trusted Firmware-M with debug symbols, set the :kconfig:option:`CONFIG_TFM_CMAKE_BUILD_TYPE_RELWITHDEBINFO` Kconfig option.
To load the Trusted Firmware-M debug symbols in the GDB tool, run the following command:

.. code-block:: console

>    (gdb) add-symbol-file build/tfm/bin/tfm_s.elf

Author: Applied edits following conversation on Teams.

You can also configure log level per logger module, for example to get more information about a given subsystem.
See :ref:`ug_logging` for details on how to enable and configure logs.

Debug libraries
---------------

The |NCS| also provides several libraries and drivers for debugging different components.
For example:

* You can use :ref:`lib_debug` for any of your applications, for example to measure CPU load or trace hardware peripheral events on pins.

* You can use :ref:`nrf_profiler` to measure performance and debug applications without introducing big performance overhead.
  This option requires introducing additional code changes: your application must register profiler events and log their occurrences.

* When working with the :ref:`SEGGER J-Link with the RTT feature <testing_rtt_connect>`, you can use the :ref:`lib_eth_rtt`, which is useful for handling data transfer.

* The Thread protocol implementation offers :ref:`pre-built libraries with debug symbols <thread_ug_feature_updating_libs>`.

* The Zigbee protocol implementation offers :ref:`lib_zigbee_osif` and :ref:`lib_zigbee_shell` with custom Kconfig options that you can set for debugging.

> 

Reviewer: There is also Memfault SDK maybe it is worth mentioning here, https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/libraries/others/memfault_ncs.html#mod-memfault.
What do you think ?

Author: Discussed adding this with @trond-snekvik and we opted to skip this mention, because it is not exactly related to debugging. It can help, but we didn't think it is something we should highlight among more debug-relevant items.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
#. Start code execution on the network core in the other debug session.

If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.

Debugging secure and non-secure firmware
========================================

When using an ``_ns`` build target, by default you can only debug firmware in the non-secure part of the application core firmware.

To debug firmware running in the secure part, you need to build Trusted Firmware-M with debug symbols enabled and load the symbols during the debugging session.
To build Trusted Firmware-M with debug symbols, set the :kconfig:option:`CONFIG_TFM_CMAKE_BUILD_TYPE_RELWITHDEBINFO` Kconfig option.
To load the Trusted Firmware-M debug symbols in the GDB tool, run the following command and answer ``y`` if prompted by the tool:

.. code-block:: console

   (gdb) add-symbol-file build/tfm/bin/tfm_s.elf

For more information about the GDB tool, see `Debug tools`_.

Debug configuration
===================

> The following Kconfig options are automatically set to ``y`` by the |VSC| debugger in the |nRFVSC| when you enable the debug configuration before building the application:

Reviewer 2: > are automatically set to ``y``

automatically.

> when you enable the debug configuration

sounds like something the user does == manually ~= automatic.

So, what is it, automatically or manual ?

> by the |VSC| debugger 

vs.
> before building the application

So is this done when building the app ?
or by the debugger, or will launching the debugger cause a rebuild of the application first.
I don't believe the debugger does anything by itself, but a rebuild could be triggered when launching the debugger.
If that's the case, then please be clear on what triggers a rebuild.

Right now, it is really unclear what happens when, and what is automatic and what is done by the user.

* :kconfig:option:`CONFIG_DEBUG_OPTIMIZATIONS` - This option limits the optimizations made by the compiler to only those that do not impact debugging.
* :kconfig:option:`CONFIG_DEBUG_THREAD_INFO` - This option adds additional information to the thread object, so that the debugger can discover the threads.
  This will work for any debugger.

> If you are programming from command line, you can set these options to ``y`` manually.

Reviewer 2: > programming from command line.

when you say programming from command line, I assume you refer to `west flash`, `nrfjprog` or similar.
How is it possible to set build options when programming ?

And what will happen if I program from command line, and afterwards launch the debugger in VSCode ?
When developing an app, those are different steps:
- compiling and linking (building)
- programming (write image to flash)
   - programming may rebuild an app if there are code changes, but this depends on user settings.
- debugging (attach debugger to running program)
  - launching a debug session may re-program the flash
  - attaching a debugger may reset the SoC

from the current text, it is not clear what each step actually does, and hence it becomes must harder to know what happens when.

   .. note::

      In the case of nRF9160 DK, the reset button must be pressed to restart the device and to start the application.

.. _gs_debugging:

Debugging an application
************************

Given that |VSC| is the main IDE for the |NCS|, set up the debug session as described in `Debugging an application`_ in the |nRFVSC| documentation.

To debug the application core firmware, you need a single debug session.
To debug the firmware running on the network core, you also need to set up a separate debug session for the application core.

Complete the following steps to start debugging the network core:

1. Set up sessions for the application core and network core as mentioned in `Debugging an application`_.

> #. Select the appropriate CPU for debugging in each session, the nRF5340 application core and the nRF5340 network core respectively.

Reviewer 3: I think we should make it generic - it was moved from nrf53 specific file so it refers to nrf53 cores

#. Once both sessions are established, execute the code on the application core.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
#. Start code execution on the network core in the other debug session.

If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.

You can also use the GDB command line tool for debugging.

.. note::

>   Debugging firmware on the application core in the non-secure domain (when using the ``nrf5340dk_nrf5340_cpuapp_ns`` build target), is currently not supported.

Reviewer 4: Probably want to generalize this as well, to something like "(when using a build target that ends in \`\`_ns\`\`)".

When debugging the network core, the application core debug session runs in the background and you can debug both cores if needed.

Complete the following steps to start debugging the network core:

1. Set up sessions for the application core and network core as mentioned in `Debugging an application`_.
#. Select the appropriate CPU for debugging in each session, corresponding to the application core and the network core of your SoC, respectively.
#. Once both sessions are established, execute the code on the application core.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
#. Start code execution on the network core in the other debug session.

If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.

.. note::
  Debugging firmware on the application core in the non-secure domain (that is, when using the ``_ns`` build target), is currently not supported.

Debugging tools
===============

> Given that |VSC| is the main IDE for the |NCS|, we recommend using the |nRFVSC| debugging tools for the |NCS|.

Author: I've rewritten this part, actually.

   .. note::

      In the case of nRF9160 DK, the reset button must be pressed to restart the device and to start the application.

.. _gs_debugging:

Debugging an application
************************

To debug the application core firmware, you need a single debug session.
Set up the debug session as described in `Debugging an application`_ in the |nRFVSC| documentation.

> To debug the firmware running on the network core, you need to set up two separate debug sessions: one for the network core and one for the application core.

Reviewer 5: Something like:
`If you use nRF53, you may want to debug both application and network core. See <link> for detailed guide.`

> To debug the firmware running on the network core, you need to set up two separate debug sessions: one for the network core and one for the application core.

Author: I'd prefer to keep everything here, not the other way around...

#. Once both sessions are established, execute the code on the application core.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
#. Start code execution on the network core in the other debug session.

If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.

.. note::
  Debugging firmware on the application core in the non-secure domain (that is, when using the ``_ns`` build target), is currently not supported.

Debugging tools
===============

For debugging with the |nRFVSC|, read `Debugging an application`_ in the |nRFVSC| documentation.
After adding the required Kconfig options to the :file:`prj.conf` file, the recommended approach is to use the `GNU Project Debugger`_ (GDB tool) for debugging.

If you are working from the command line, you can use west with the GDB tool.
For details, read the :ref:`Debugging with west debug <zephyr:west-debugging>` section on the :ref:`zephyr:west-build-flash-debug` page in the Zephyr documentation.

> A useful tool for debugging mesh networking protocols, such as :ref:`ug_thread` or :ref:`ug_zigbee`, is the `nRF Sniffer for 802.15.4`_.

Reviewer 5: btw. nRF Sniffer is also very useful for Bluetooth communication

Reviewer 5: IF we remove GDB information from this section we can name it e.g. `Sniffing the communication`

Debug logging
=============

The |NCS| provides a logging system that offers a variety of options for several logging backends.
For more information, read the :ref:`ug_logging` user guide.

Custom debugging configuration
==============================

The following Kconfig options are used by the |VSC| debugger in the |nRFVSC|:

* :kconfig:option:`CONFIG_DEBUG_OPTIMIZATIONS` set to ``y``

> * :kconfig:option:`CONFIG_DEBUG_THREAD_INFO` set to ``y``

Reviewer 5: Maybe we could add a separate section that lists all of the useful debug options? We could add these options and options mentioned in `Debugging an application`_ in the |nRFVSC| documentation there.
The section could be named `Debugging configuration`, we could place it above `Debugging tools`.

Author: Wouldn't adding a link to the Kconfig search with "DEBUG" be better instead?

Reviewer 5: I think not, because it would add quite a lot of results. I would only put the most commonly used ones here to avoid flooding customers wtih information.

Some applications and samples may require additional Kconfig options to be set when debugging.
For example, nRF9160's :ref:`nRF Cloud MQTT multi-service <nrf_cloud_mqtt_multi_service>` sample requires enabling :ref:`CONFIG_MQTT_MULTI_SERVICE_LOG_LEVEL_DBG <CONFIG_MQTT_MULTI_SERVICE_LOG_LEVEL_DBG>`.

> Check the documentation of the application, sample, and protocol to see if it offers custom debugging settings.

Reviewer 5: Maybe we can move it under debug logging and make it an example of configuring log level. I think setting this may not be mandatory in every use-case. It's just useful if you want to see more detailed logs from a given logger module.

> Other applications and samples may require using a specific build type file when :ref:`configuring the build settings <gs_modifying_build_types>`.

Reviewer 5: `Some applications and samples provide a specific build type that enables additional debug functionalities.`
`See :ref:`configuring the build settings <gs_modifying_build_types>`

Custom libraries and drivers for debugging
==========================================

> 

Reviewer 5: Maybe we could also add a sentence of introduction and split it into sections to avoid putting everything together (the libraries are not related, the only common part is the fact that they are available in the NCS and they are used for debugging purposes)

   .. note::

      In the case of nRF9160 DK, the reset button must be pressed to restart the device and to start the application.

.. _gs_debugging:

Debugging an application
************************

To debug an application, set up the debug session as described in `Debugging an application`_ in the |nRFVSC| documentation.

> If you use a multi-core SoC, for example from the nRF53 Series, you need a single debug session to debug the application core firmware.

Reviewer 4: Does SoC need to be opened as System on Chip? Or do we consider the abbreviation common enough?

Author: I think it's common enough. We also have it in the glossary.

Complete the following steps to start debugging the network core:

1. Set up sessions for the application core and network core as mentioned in `Debugging an application`_.
#. Select the appropriate CPU for debugging in each session, corresponding to the application core and the network core of your SoC, respectively.
#. Once both sessions are established, execute the code on the application core.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
#. Start code execution on the network core in the other debug session.

If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.

.. note::
  Debugging firmware on the application core in the non-secure domain (that is, when using the ``_ns`` build target), is currently not supported.

Debug configuration
===================

The following Kconfig options are used by the |VSC| debugger in the |nRFVSC|:

* :kconfig:option:`CONFIG_DEBUG_OPTIMIZATIONS` set to ``y``
* :kconfig:option:`CONFIG_DEBUG_THREAD_INFO` set to ``y``

Debug logging
-------------

> Some protocols, applications, and samples may require additional Kconfig options to be set when debugging.

Reviewer 5: Maybe we could rephrase it a bit:
```
You can use logging system to get more information about the state of your application.
Logs are integrated to many various modules and subsystems in the nRF Connect SDK and Zephyr.
These logs are visible once you configure logger for your application.
You can also configure log level per logger module to, for example, get more information about given subsystem.
See :ref:ug_logging` for details on how to enable and configure logs.
```

Among these are logging backend configuration options that need to be set for the logging system to provide debug logs.

> For example, nRF9160's :ref:`nRF Cloud MQTT multi-service <nrf_cloud_mqtt_multi_service>` sample will start printing debug logs after you enable :ref:`CONFIG_MQTT_MULTI_SERVICE_LOG_LEVEL_DBG <CONFIG_MQTT_MULTI_SERVICE_LOG_LEVEL_DBG>`.

Reviewer 4: ```suggestion
For example, the :ref:`nRF Cloud MQTT multi-service <nrf_cloud_mqtt_multi_service>` nRF9160 sample will start printing debug logs after you enable :ref:`CONFIG_MQTT_MULTI_SERVICE_LOG_LEVEL_DBG <CONFIG_MQTT_MULTI_SERVICE_LOG_LEVEL_DBG>`.
```
This could also work with just "the :ref:\`nrf_cloud_mqtt_multi_service\` sample", I think.

Author: Removed this line.

For more information, read the :ref:`ug_logging` user guide.

> To list all debug-related Kconfig options in the |NCS|, search for "DEBUG" on the :ref:`kconfig-search` page.

Reviewer 5: I think the list may not be complete here, I would remove this information, because it could be confusing.

Debug build types
-----------------

Some applications and samples provide a specific build type that enables additional debug functionalities.
You can select these build types when you are :ref:`configuring the build settings <gs_modifying_build_types>`.

Debug libraries
---------------

The |NCS| also provides several libraries and drivers for debugging different components.

> * When working with the :ref:`SEGGER J-Link with the RTT feature <testing_rtt_connect>`, you can use the :ref:`lib_eth_rtt`, useful to handle data transfer.

Reviewer 5: Maybe we could also mention nRF Profiler here (it's great tool to measure performance and debug applications without introducing big performance overhead)? We could even make a separate section about it.

> * When working with the :ref:`SEGGER J-Link with the RTT feature <testing_rtt_connect>`, you can use the :ref:`lib_eth_rtt`, useful to handle data transfer.

Reviewer 4: ```suggestion
* When working with the :ref:`SEGGER J-Link with the RTT feature <testing_rtt_connect>`, you can use the :ref:`lib_eth_rtt` to handle data transfer.
```
or
```suggestion
* When working with the :ref:`SEGGER J-Link with the RTT feature <testing_rtt_connect>`, you can use the :ref:`lib_eth_rtt`, which is useful for handling data transfer.
```

* The Thread protocol offers :ref:`pre-built libraries with debug symbols <thread_ug_feature_updating_libs>`.

* The Zigbee protocol offers :ref:`lib_zigbee_osif` and :ref:`lib_zigbee_shell` with custom Kconfig options that you can set for debugging.

Debug tools
===========

The main recommended tool for debugging in the |NCS| is the `GNU Project Debugger`_ (GDB tool).

* When working with the |nRFVSC|, use the GDB tool after adding the required Kconfig options to the :file:`prj.conf` file.
* If you are working from the command line, you can use west with the GDB tool.
  For details, read the :ref:`Debugging with west debug <zephyr:west-debugging>` section on the :ref:`zephyr:west-build-flash-debug` page in the Zephyr documentation.

> A useful tool for debugging the communication over Bluetooth and mesh networking protocols, such as :ref:`ug_thread` or :ref:`ug_zigbee`, is the `nRF Sniffer for 802.15.4`_.

Reviewer 5: `The nRF Sniffer allows you to look into data exchanged over the air between devices.`

   .. note::

      In the case of nRF9160 DK, the reset button must be pressed to restart the device and to start the application.

.. _gs_debugging:

Debugging an application
************************

To debug an application, set up the debug session as described in `Debugging an application`_ in the |nRFVSC| documentation.

> If you use a multi-core SoC, for example from the nRF53 Series, you need a single debug session to debug the application core firmware.
> To debug the firmware running on the network core, you need to set up two separate debug sessions: one for the network core and one for the application core.

Reviewer 2: First it is mentioned that a single debug session is needed for debugging the app core is described, but right after it is mentioned that two separate debug sessions are needed for debugging both cores.

I guess that actually means that the first line is only valid for only app core debugging.
So how about:
```suggestion
If you use a multi-core SoC, for example from the nRF53 Series, and you only wish to debug the application core firmware then a single debug session is sufficient.
To debug the firmware running on the network core, you need to set up two separate debug sessions: one for the network core and one for the application core.
```

When debugging the network core, the application core debug session runs in the background and you can debug both cores if needed.

Complete the following steps to start debugging the network core:

1. Set up sessions for the application core and network core as mentioned in `Debugging an application`_.
#. Select the appropriate CPU for debugging in each session, corresponding to the application core and the network core of your SoC, respectively.
#. Once both sessions are established, execute the code on the application core.

   The startup code releases the ``NETWORK.FORCEOFF`` signal to start the network core and allocates the necessary GPIO pins for it.
#. Start code execution on the network core in the other debug session.

If you want to reset the network core while debugging, make sure to first reset the application core and execute the code.

.. note::

>   Debugging firmware on the application core in the non-secure domain (that is, when using an ``_ns`` build target), is currently not supported.

Author: Implemented! Thanks for help!

Debug configuration
===================

The following Kconfig options are used by the |VSC| debugger in the |nRFVSC|:

> * :kconfig:option:`CONFIG_DEBUG_OPTIMIZATIONS` set to ``y``

Reviewer 3: I have not used this VSC debugger but does it really enable this config options automatically? That would be problematic... Adding DEBUG_OPTIMIZATIONS is needed only if you want to have correct C code stepping and you don't care about it if you are fine with assembler.
Forcefully enabling that might impact application sizes and cause some apps/samples not to build.

Or is it that it is suggested to enable these options to have better debugging experience?

Author: @trond-snekvik , how does this look from the extension perspective? The VSC ext docs don't state these are enabled by default, but they also use wording that might be ambiguous ("**If you forgot** to enable debugging during the build configuration step, add the following lines into the application configuration file [...]").

Reviewer 2: I hope this is just bad wording in the guide, but until we know what is correct could @pdunaj block the PR, so that it doesn't get merged until this is resolved ?

Author: Put DNM on the PR for now :) I'll check with Trond today.

Reviewer 2: Is `Enable debug configuration` a setting that the user selects independently of launching the debugger itself ?
Or is it automatically selected by the debugger.

I hope it's an independent setting that user can select independently of the debugger.
If that's the case, then the wording should be updated to reflect that selection of the setting could improve the debugger session, but that it is not required for debugging.

Reviewer 3: As @tejlmand mentioned, it's not a problem as long as user can control this. I assume this wording will be fixed.

Author: The user can indeed enable or disable this manually, yes.

Reviewer 2: > (We're planning to update the wording in this task for the VSC extension: https://projecttools.nordicsemi.no/jira/browse/VSC-978 )

I see no mentioning of this PR or updating of the wording, so until it is clear in the task what must be done wrt. wording then I cannot give a +1, cause there's too high a risk that the wording will not be updated if not explicitly stated.

For details, see the respective documentation pages of the modules.

Debug build types
-----------------

Some applications and samples provide a specific build type that enables additional debug functionalities.
You can select build types when you are :ref:`configuring the build settings <gs_modifying_build_types>`.

Debug logging
-------------

You can use logging system to get more information about the state of your application.
Logs are integrated to many various modules and subsystems in the |NCS| and Zephyr.
These logs are visible once you configure logger for your application.

You can also configure log level per logger module, for example to get more information about a given subsystem.
See :ref:`ug_logging` for details on how to enable and configure logs.

Debug libraries
---------------

The |NCS| also provides several libraries and drivers for debugging different components.
For example:

* You can use :ref:`lib_debug` for any of your applications, for example to measure CPU load or trace hardware peripheral events on pins.

* You can use :ref:`nrf_profiler` to measure performance and debug applications without introducing big performance overhead.

>   This option requires introducing additional code changes to your application.

Reviewer 5: Maybe we could add that:
`Your application must register profiler events and log their occurrences.`

Reviewer 5: To clarify what is needed

* When working with the :ref:`SEGGER J-Link with the RTT feature <testing_rtt_connect>`, you can use the :ref:`lib_eth_rtt`, which is useful for handling data transfer.

> * The Thread protocol offers :ref:`pre-built libraries with debug symbols <thread_ug_feature_updating_libs>`.

Reviewer 5: protocol or implementation (same with Zigbee)?

Author: Actually both!

* The Zigbee protocol offers :ref:`lib_zigbee_osif` and :ref:`lib_zigbee_shell` with custom Kconfig options that you can set for debugging.

Debugging stack overflows
-------------------------

> One of the potential root causes of fatal errors in the application are stack overflows.

Reviewer 4: ```suggestion
One of the potential root causes of fatal errors in an application are stack overflows.
```
Since we aren't talking about a specific application.

Author: Ahh, articles, the bane of article-free language speakers...

Reviewer 4: I'll drink to that (after hours).
